tests/sim_trace_viz.py

```diff
@@ -0,0 +1,129 @@
+    "recv_propose_request_vote": "RPRv",
+    "become_candidate": "BCan",
+    "BecomeLeader": "BLea",
+    "become_follower": "BFol",
+    "AdvanceCommitIndex": "Cmt",
+    "bootstrap": "Boot",
+    None: "",
+    "SignCommittableMessages": "Sig",
+    "SIMTimeout": "Tmou",
+    "RcvUpdateTerm": "UTrm",
+    "SIMCheckQuorum": "ChkQ"
+}
+
+TAG = {"Y": ":white_check_mark:", "N": ":x:", " ": "  ", "S": ":pencil:"}
+
+
+def digits(value):
+    return len(str(value))
+
+
+def diffed_key(old, new, key, suffix, size):
+    if old is None or old[key] == new[key]:
+        return f"{new[key]:>{size}}{suffix}"
+    color = "bright_white on red"
+    return f"[{color}]{new[key]:>{size}}{suffix}[/{color}]"
+
+
+def render_state(state, func, old_state, tag, cfg):
+    if state is None:
+        return " "
+    ls = LEADERSHIP_STATUS[state["leadership_state"]]
+    nid = state["node_id"]
+    v = diffed_key(old_state, state, "current_view", "", cfg.view)
+    i = diffed_key(old_state, state, "last_idx", "", cfg.index)
+    c = diffed_key(old_state, state, "commit_idx", "", cfg.commit)
+    f = FUNCTIONS[func]
+    opc = "bold bright_white on red" if func else "normal"
+    # import pdb; pdb.set_trace()
+    return f"[{opc}]{nid:>{cfg.nodes}}{ls}{f:<4} [/{opc}]{TAG[tag]} {v}.{i} {c}"
+
+
+class DigitsCfg:
+    nodes = 0
+    view = 0
+    index = 0
+    commit = 0
+    ts = 0
+
+def extract_node_state(global_state, node_id):
+    return {
+        "node_id": node_id,
+        "leadership_state": global_state["leadershipState"][node_id],
+        "current_view": global_state["currentTerm"][node_id],
+        "last_idx": len(global_state["log"][node_id]),
+        "commit_idx": global_state["commitIndex"][node_id],
+    }
+
+def table(entries):
+    nodes = []
+    max_view = 0
+    max_index = 0
+    max_commit = 0
+    max_ts = 0
+    rows = []
+    for pre, action, post in entries:
+        node_id = action["context"]["i"]
+        if node_id not in nodes:
+            nodes.append(node_id)
+        max_view = max(max_view, post[1]["currentTerm"][node_id])
+        max_index = max(max_index, len(post[1]["log"][node_id]))
+        max_commit = max(max_commit, post[1]["commitIndex"][node_id])
+        max_ts += 1
+    dcfg = DigitsCfg()
+    dcfg.nodes = len(max(nodes, key=len))
+    dcfg.view = digits(max_view)
+    dcfg.index = digits(max_index)
+    dcfg.commit = digits(max_commit)
+    dcfg.ts = digits(max_ts)
+    ts = 0
+    for pre, action, post in entries:
+        ts += 1
+        node_id = action["context"]["i"]
+        # TODO: need success, vote granted and globally committable
+        # to be extracted from pre or post state
+        states = [
+            (
+                extract_node_state(post[1], node),
+                action["name"] if node == node_id else None,
+                extract_node_state(pre[1], node) if node == node_id else None,
+                " " # tag
+            )
+            for node in nodes
```

Comment:
```suggestion
            for node in sorted(nodes)
```
<img width="435" alt="image" src="https://github.com/microsoft/CCF/assets/6000239/4697ae3b-6857-4d86-943a-695545757455">

rather than
<img width="437" alt="image" src="https://github.com/microsoft/CCF/assets/6000239/e9f67f4f-4f66-443c-9731-1f14fcb53590">

Reply:
Good point, in the impl one, the order deliberately is order of appearance in the trace, but here relying on the names is preferable.